File: frontend/schemas/customer.ts
import { z } from "zod";

// Mirrors backend/app/models/enums.py::CustomerStatus
export const customerStatuses = ["onboarding", "active", "at_risk", "churned"] as const;
export type CustomerStatusValue = (typeof customerStatuses)[number];

export const customerSorts = ["created_at", "name", "health_score", "arr"] as const;
export type CustomerSortValue = (typeof customerSorts)[number];

// Mirrors backend/app/schemas/customer.py::CustomerBase/CustomerCreate. `arr`
// is a Decimal on the backend (serialized as a JSON string, see
// app/schemas/customer.py::CustomerBase.arr) — the form keeps it a string
// and only shapes it (ge 0, <=2 decimal places) rather than coercing to
// number, so "12.5" round-trips without float rounding surprises.
export const customerFormSchema = z.object({
  name: z.string().min(1).max(160),
  company: z.string().min(1).max(160),
  email: z.email(),
  phone: z.string().max(32).optional().or(z.literal("")),
  industry: z.string().max(80).optional().or(z.literal("")),
  status: z.enum(customerStatuses),
  health_score: z.coerce.number().int().min(0).max(100),
  arr: z
    .string()
    .regex(/^\d{1,10}(\.\d{1,2})?$/, "Enter a number with up to 2 decimal places")
    .optional()
    .or(z.literal("")),
  // Only rendered for admin/manager; a csm never sees this field, so the
  // backend's forced-self-ownership rule is never exercised through it.
  owner_id: z.uuid().optional().or(z.literal("")),
});
// Output (post-coercion, e.g. health_score: number) vs. input (pre-coercion,
// what raw <input> values look like) — z.coerce fields make these diverge,
// so useForm is typed with both (see CustomerForm.tsx).
export type CustomerFormInput = z.infer<typeof customerFormSchema>;
export type CustomerFormRawInput = z.input<typeof customerFormSchema>;

File: backend/app/schemas/customer.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator

from app.models.enums import CustomerStatus
from app.schemas.user import UserOut

CustomerSort = Literal["created_at", "name", "health_score", "arr"]
SortOrder = Literal["asc", "desc"]


def _lower(v: str) -> str:
    return v.strip().lower()


class CustomerBase(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    company: str = Field(min_length=1, max_length=160)
    email: EmailStr
    phone: str | None = Field(default=None, max_length=32)
    industry: str | None = Field(default=None, max_length=80)
    status: CustomerStatus = CustomerStatus.onboarding
    health_score: int = Field(default=50, ge=0, le=100)
    arr: Decimal | None = Field(default=None, ge=0, max_digits=12, decimal_places=2)

    _lower_email = field_validator("email", mode="after")(_lower)


class CustomerCreate(CustomerBase):
    # Service decides: csm -> forced self; admin/manager -> self if omitted, else this value.
    owner_id: uuid.UUID | None = None


class CustomerUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=160)
    company: str | None = Field(default=None, min_length=1, max_length=160)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=32)
    industry: str | None = Field(default=None, max_length=80)
    status: CustomerStatus | None = None
    health_score: int | None = Field(default=None, ge=0, le=100)
    arr: Decimal | None = Field(default=None, ge=0, max_digits=12, decimal_places=2)
    owner_id: uuid.UUID | None = None

    _lower_email = field_validator("email", mode="after")(_lower)


class CustomerListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    company: str
    email: str
    industry: str | None
    status: CustomerStatus
    health_score: int
    arr: Decimal | None
    owner_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class CustomerOut(CustomerListItem):
    phone: str | None
    interaction_count: int
    owner: UserOut | None = None
